chirho/dynamical/internals/backends/torchdiffeq.py

```diff
@@ -102,10 +99,10 @@ def _batched_odeint(
     return yt if event_fn is None else (event_t, yt)
 
 
-@ode_simulate.register(TorchDiffEq)
+@_simulate.register(ODEDynamics, TorchDiffEq)
```

Comment:
`multimethod` dispatches on all of the arguments to `_simulate`, so what does this registration mean?

Reply:
My understanding is that this registration means that we dispatch only on the first two arguments of `_simulate`, and not the remaining. This isn't documented in `multimethod` from what I could find, when I remove remove the arguments I get the default `NotImplementedError` for `_simulate` or the other dispatches.

Reply:
I don't know if that's precisely true (I would guess it implicitly provides a type for the remaining arguments, maybe `object` or whatever is in the signature?), but I don't think it's a good idea to rely on undocumented behavior in upstream dependencies since it can change without warning in ways that are difficult or impossible to work around.

Reply:
Good point. I'll try and debug why the default dispatch on the full collection of arguments isn't working as expected.

Reply:
@eb8680 , after a bit of digging around I've found two options that rely less on undocumented behavior and pass tests. I have not been able to get `multimethod` to work with parametric types fully specified.

Option 1: Pass unparameterized types to the `*args` of `register` and leave fully parametric types explicit in the function signature. E.g.

```
@_simulate.register(ODEDynamics, TorchDiffEq, State, torch.Tensor, torch.Tensor)
def torchdiffeq_ode_simulate(
    dynamics: ODEDynamics[torch.Tensor, torch.Tensor],
    solver: TorchDiffEq,
    initial_state: State[torch.Tensor],
    start_time: torch.Tensor,
    end_time: torch.Tensor,
) -> State[torch.Tensor]:
    timespan = torch.stack((start_time, end_time))
    trajectory = _torchdiffeq_ode_simulate_inner(
        dynamics, initial_state, timespan, **solver.odeint_kwargs
    )
    return trajectory[..., -1].to_state()
```

Option 2: Remove type parameters from function signature and use default `register` without any `*args`. E.g.


```
@_simulate.register
def torchdiffeq_ode_simulate(
    dynamics: ODEDynamics,
    solver: TorchDiffEq,
    initial_state: State,
    start_time: torch.Tensor,
    end_time: torch.Tensor,
) -> State:
    timespan = torch.stack((start_time, end_time))
    trajectory = _torchdiffeq_ode_simulate_inner(
        dynamics, initial_state, timespan, **solver.odeint_kwargs
    )
    return trajectory[..., -1].to_state()
```

Which of these two would you prefer? Alternatively, would you like something else?

Unrelated: I think there is some misuse of types a bit scattered throughout the module, so a separate type refactoring PR (ideally pair programmed) might be nice before merging into master.